### PR TITLE
Dpp 1558 datalock rework

### PR DESCRIPTION
**Pull request by carlbm**

### CODE REVIEW

File: src/AcceptanceTesting/Features/additional_payments.feature [hunk 1]
```diff
@@ -1108,7 +1107,7 @@ Scenario:658-AC01 DAS learner, takes an English qualification that has a planned
 	And the following commitments exist:
 		| commitment Id | version Id | ULN       | start date | end date   | agreed price | status    | effective from | effective to |
 		| 1             | 1-001      | learner a | 01/08/2017 | 01/08/2018 | 15000        | active    | 01/08/2017     | 10/06/2018   |
-		| 1             | 2-001      | learner a | 01/08/2017 | 01/08/2018 | 14000        | cancelled | 11/06/2018     |              |
+		| 1             | 2-001      | learner a | 01/08/2017 | 01/08/2018 | 14000        | active    | 11/06/2018     |              |
```
Comment (Pauljgraham) [inline src/AcceptanceTesting/Features/additional_payments.feature:1110]: doesn't changing this to 'active' change the Test results?

File: src/AcceptanceTesting/Features/employer_stops_payments.feature [hunk 2]
```diff
@@ -27,8 +27,8 @@ Feature: Employer stops payments on a commitment
         Given levy balance > agreed price for all months
         And the following commitments exist:
             | commitment Id | version Id | ULN       | start date | end date   | status | agreed price | effective from | effective to |
-            | 1             | 1          | learner a | 01/09/2017 | 08/09/2018 | active | 15000        | 01/09/2017     | 31/08/2017   |
-            | 1             | 2          | learner a | 01/09/2017 | 08/09/2018 | paused | 15000        | 01/09/2017     |              |
+            | 1             | 1          | learner a | 01/09/2017 | 08/09/2018 | active | 15000        | 01/09/2017     | 02/09/2017   |
+            | 1             | 2          | learner a | 01/09/2017 | 08/09/2018 | paused | 15000        | 02/09/2017     |              |
```
Comment (Pauljgraham) [inline src/AcceptanceTesting/Features/employer_stops_payments.feature:31]: Can we discuss these AT changes with Andy?

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs [hunk 3]
```diff
@@ -192,27 +204,65 @@ private static CommitmentReferenceData ParseCommitmentsTableRow(TableRow row, Co
...
+
             if (effectiveFrom == null)
             {
                 effectiveFrom = startDate;
             }
 
+            if (frameworkCode > 0)
+            {
+                standardCode = 0;
+            }
```
Comment (Pauljgraham) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:271]: Why are we introducing this default?

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs [hunk 4]
```diff
@@ -192,27 +204,65 @@ private static CommitmentReferenceData ParseCommitmentsTableRow(TableRow row, Co
...
+
+            if (standardCode > 0)
+            {
+                programmeType = 25;
```
Comment (Pauljgraham) [inline src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:275]: and this?

File: src/PeriodEnd/PaymentsDue/DeployPaymentsDue/copy mappings/DasPaymentsDueILRCopyToTransientMapping.xml [hunk 5]
```diff
@@ -0,0 +1,281 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (Pauljgraham) [inline src/PeriodEnd/PaymentsDue/DeployPaymentsDue/copy mappings/DasPaymentsDueILRCopyToTransientMapping.xml:1]: Why have we used a new file/?

File: src/PeriodEnd/PaymentsDue/DeployPaymentsDue/sql/ddl/Ilr.Bulk.PaymentsDue.DDL.Views.sql [hunk 6]
```diff
@@ -0,0 +1,197 @@
+﻿-----------------------------------------------------------------------------------------------------------------------------------------------
```
Comment (Pauljgraham) [inline src/PeriodEnd/PaymentsDue/DeployPaymentsDue/sql/ddl/Ilr.Bulk.PaymentsDue.DDL.Views.sql]: Is this a new file?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.IntTests/Infrastructure/Data/Repositories/Temp.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using NUnit.Framework;
+
+namespace SFA.DAS.CollectionEarnings.DataLock.IntegrationTests.Infrastructure.Data.Repositories
+{
+    [TestFixture]
+    public class Temp
+    {
+        [Test]
+        public void IsNotATest()
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.IntTests/Infrastructure/Data/Repositories/Temp.cs:9]: Why is this here?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.IntTests/packages.config [hunk 8]
```diff
@@ -1,9 +1,15 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="AutoFixture" version="4.5.0" targetFramework="net462" />
+  <package id="AutoFixture.AutoMoq" version="4.5.0" targetFramework="net462" />
+  <package id="AutoFixture.NUnit3" version="4.5.0" targetFramework="net462" />
   <package id="Castle.Core" version="4.2.1" targetFramework="net45" />
-  <package id="Dapper" version="1.50.2" targetFramework="net45" />
+  <package id="Dapper" version="1.50.2" targetFramework="net45" requireReinstallation="true" />
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.IntTests/packages.config:7]: why does it require re-installation?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.UnitTests/Application/DataLock/Matcher/UkprnMatchHandler/Match/WhenCalled.cs [hunk 9]
```diff
@@ -90,13 +93,13 @@ public void ThenNextHanderShouldNotBeCalled()
 
 
             // Act
-            var matchResult = _matcher.Match(commitments, priceEpisode,accounts);
+            var matchResult = _matcher.Match((IReadOnlyList<Commitment>) commitments, priceEpisode, (DateTime)new DateTime());
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.UnitTests/Application/DataLock/Matcher/UkprnMatchHandler/Match/WhenCalled.cs]: why do we need to cast 'new DateTime()' and also what date does new DateTime() give us?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/PausedCommitmentMatcher.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.CollectionEarnings.DataLock.Application.DataLock.Matcher
+{
+    class PausedCommitmentMatcher : MatchHandler
+    {
+        public PausedCommitmentMatcher(MatchHandler nextMatchHandler) : base(nextMatchHandler)
+        {}
+
+        public override bool StopOnError { get { return false; } }
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/PausedCommitmentMatcher.cs:15]: I don't think you need to override this. As it's default behaviour

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs [hunk 11]
```diff
@@ -39,30 +46,32 @@ public virtual void Process()
...
-                        WriteDataLockValidationErrorsOrThrow(dataLockValidationResult);
-                        WriteDataLockPriceEpisodeMatches(dataLockValidationResult);
-                        WriteDataLockPriceEpisodePeriodMatches(dataLockValidationResult);
-                    }
-                    else
-                    {
-                        _logger.Info("No price episodes found.");
-                    }
+                    var commitments = CommitmentsForProvider(provider.Ukprn);
+                    var providerCommitments = new ProviderCommitments(commitments);
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs]: Could these be combined to call a single private function CreateProviderCommitments(ukprn) as commitments is never used

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs [hunk 12]
```diff
@@ -98,153 +107,22 @@ private GetDasAccountsQueryResponse ReturnValidGetDasAccountsQueryResponseOrThro
...
-                _mediator.Send(new GetProviderCommitmentsQueryRequest
-                {
-                    Ukprn = ukprn
-                });
-
-            if (!commitments.IsValid)
-            {
-                throw new DataLockException(DataLockException.ErrorReadingCommitmentsMessage, commitments.Exception);
-            }
+            var commitments = _commitmentRepository.GetProviderCommitments(ukprn);
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs]: Are we sure want to introduce a different style here. I totally agree the Mediator pattern adds an additional layer here but in theory it should protect the domain from the use of database Entities. Although  our apps do not behave in a very consistent way

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs [hunk 13]
```diff
@@ -98,153 +107,22 @@ private GetDasAccountsQueryResponse ReturnValidGetDasAccountsQueryResponseOrThro
...
-            var priceEpisodesQueryResponse = _mediator.Send(new GetProviderPriceEpisodesQueryRequest
-            {
-                Ukprn = ukprn
-            });
-
-            if (!priceEpisodesQueryResponse.IsValid)
-            {
-                throw new DataLockException(DataLockException.ErrorReadingPriceEpisodesMessage, priceEpisodesQueryResponse.Exception);
-            }
+            var earnings = _rawEarningsRepository.GetAllAct1ForProvider(ukprn);
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs]: ditto

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs [hunk 14]
```diff
@@ -39,30 +46,32 @@ public virtual void Process()
 
             var providersQueryResponse = ReturnValidGetProvidersQueryResponseOrThrow();
 
-            var dasAccountsQueryResponse = ReturnValidGetDasAccountsQueryResponseOrThrow();
+            var dasAccountsQueryResponse = ReturnValidGetDasAccountsQueryResponseOrThrow().Items;
+            var dasAccountIdsThatHaveNonPayableFlagSet = ImmutableHashSet
+                .Create(dasAccountsQueryResponse
+                    .Where(x => x.IsLevyPayer == false)
+                    .Select(x => x.AccountId)
+                    .ToArray());
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/DataLockProcessor.cs]: Could this be combined into a single helper function. Ideally it could be a separate service. Perhaps called GetAccountsThatHave....

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+                    break;
+            }
+
+            return date;
+        }
+
+        public void ValidateInitialResult(RawEarning earning, List<string> errors, TransactionTypesFlag paymentType,
+            List<Commitment> commitments, DatalockValidationResult result, ImmutableHashSet<long> accountsWithNonPayableFlagSet,
+            List<Commitment> allCommitments, DateTime censusDate)
+        {
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs:116]: It's quite unclear what this function is actually doing? It's either updating the errors parameter or adding a record to DataLockValidationResult parameter.

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+
+            result.AddDistinctRecords(earning, errors, paymentType, commitments.First());
+        }
+
+        private void CheckForEarlierStartDate(RawEarning earning, List<Commitment> commitments, DateTime censusDate, TransactionTypesFlag paymentType, DatalockValidationResult result)
+        {
+            var matchResult = _datalockMatcher.Match(commitments, earning, censusDate);
+            if (!matchResult.ErrorCodes.Any() && matchResult.Commitments.Any(x => x.Ukprn == earning.Ukprn))
+            {
+                if (commitments.Any(x => x.PausedOnDate.HasValue || x.PaymentStatus == 2))
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: Do we really want to use payment status literals?

Also we calling the Match again at this lower level, I can see why, but it's sort of hidden away.

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+            var month = episodeStartDate.Month;
+            if (month < 8)
+            {
+                return new DateTime(episodeStartDate.Year - 1, 8, 1);
+            }
+            return new DateTime(episodeStartDate.Year, 8, 1);
+        }
+    }
+
+    public class DatalockValidationResult
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: A quick question of styles. If a class is public do we not have it in it's own file. The same with Interfaces etc?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+                            learnerCommitments, result, TransactionTypesFlag.SecondEmployerProviderIncentives);
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private void ProcessEarning(ImmutableHashSet<long> accountsWithNonPayableFlagSet, RawEarning earning,
+            LearnerCommitments learnerCommitments, DatalockValidationResult result, TransactionTypesFlag earningType)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs:82]: It's unclear what this function is doing by its name

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+            {
+                errors.Add(DataLockErrorCodes.MultipleMatches);
+            }
+            else if (commitments.Count == 0)
+            {
+                CheckForEarlierStartDate(earning, allCommitments, censusDate, paymentType, result);
+                return;
+            }
+
+            result.AddDistinctRecords(earning, errors, paymentType, commitments.First());
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: Again the function name here is very general and not very descriptive

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+            if (payable)
+            {
+                var validationErrorsToRemove = ValidationErrors.Where(x =>
+                        x.Ukprn == earning.Ukprn &&
+                        x.LearnRefNumber == earning.LearnRefNumber &&
+                        x.PriceEpisodeIdentifier == earning.PriceEpisodeIdentifier)
+                    .ToList();
+
+                foreach (var validationError in validationErrorsToRemove)
+                {
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: Ncrunch is telling me this is never tested

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+                        PriceEpisodePeriodMatches.FirstOrDefault(x => 
+                            x.AimSeqNumber == earning.AimSeqNumber &&
+                            x.LearnRefNumber == earning.LearnRefNumber &&
+                            x.PriceEpisodeIdentifier == earning.PriceEpisodeIdentifier &&
+                            x.Ukprn == earning.Ukprn && 
+                            x.CommitmentId == commitment.CommitmentId &&
+                            x.Period == earning.Period &&
+                            x.TransactionTypesFlag == paymentType &&
+                            x.Payable &&
+                            x.VersionId == commitment.VersionId) == null)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: This is a very verbose if clause, it looks more like a SQL statement and it would be more readable if it used some methods instead of doing everything at once

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+
+    public class DatalockValidationResult
+    {
+        public List<DatalockValidationError> ValidationErrors { get; } = new List<DatalockValidationError>();
+        public List<DatalockValidationErrorByPeriod> ValidationErrorsByPeriod { get; } = new List<DatalockValidationErrorByPeriod>();
+        public List<PriceEpisodePeriodMatchEntity> PriceEpisodePeriodMatches { get; } = new List<PriceEpisodePeriodMatchEntity>();
+        public List<PriceEpisodeMatchEntity> PriceEpisodeMatches { get; } = new List<PriceEpisodeMatchEntity>();
+        public List<DatalockOutputEntity> DatalockOutputEntities { get; } = new List<DatalockOutputEntity>();
+        
+        public void AddDistinctRecords(RawEarning earning, List<string> errors, TransactionTypesFlag paymentType, CommitmentEntity commitment)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: Not a great name

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+                    }
+
+                    if (ValidationErrorsByPeriod.FirstOrDefault(x =>
+                            x.LearnRefNumber == earning.LearnRefNumber &&
+                            x.AimSeqNumber == earning.AimSeqNumber &&
+                            x.PriceEpisodeIdentifier == earning.PriceEpisodeIdentifier &&
+                            x.RuleId == error &&
+                            x.Period == earning.Period &&
+                            x.Ukprn == earning.Ukprn) == null)
+                    {
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: this looks similar or the same as the statement above. So a private or extension method might be clearer

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+                        ValidationErrorsByPeriod.Add(new DatalockValidationErrorByPeriod()
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: () aren't needed

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 25]
```diff
@@ -0,0 +1,295 @@
...
+                    CommitmentId = commitment.CommitmentId,
+                    IsSuccess = payable,
+                    LearnRefNumber = earning.LearnRefNumber,
+                    PriceEpisodeIdentifier = earning.PriceEpisodeIdentifier,
+                    Ukprn = earning.Ukprn,
+                });
+            }
+        }
+    }
+}
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: It's a very big method and it could be simplified by splitting out some of the code.

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+using System.Linq;
+using SFA.DAS.CollectionEarnings.DataLock.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.CollectionEarnings.DataLock.Domain
+{
+    public class LearnerCommitments
+    {
+        public List<Commitment> Commitments { get; }
+        
+        public LearnerCommitments(long key, IEnumerable<CommitmentEntity> commitments)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs]: key is not used, either remove it or add it as a property. But should be called Uln

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+            }
+
+            var lastCommitment = Commitments
+                .OrderByDescending(x => x.EffectiveStartDate)
+                .ThenByDescending(x => x.CommitmentId)
+                .FirstOrDefault();
+            if (lastCommitment != null)
+            {
+                lastCommitment.EffectiveEndDate = null;
+            }
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs:54]: this is unclear, there is implied logic to set EffectiveEndDate to Null, but at first view it looks odd. If you are setting the last commitment EffectiveDate property, we should have a private static function doing this.

I'm not a great fan of putting this much logic in a constructor. I think this type f logic can go into a helper or service. But this is a hangup I have from previous languages

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+                                              (x.PaymentStatus == 3 &&
+                                               x.WithdrawnOnDate >= date)))
+                .ToList();
+        }
+
+        public IReadOnlyList<CommitmentEntity> NonActiveCommitmentsForDate(DateTime date)
+        {
+            return Commitments.Where(x => x.EffectiveStartDate < date &&
+                                          (x.WithdrawnOnDate.HasValue ||
+                                          x.PausedOnDate.HasValue))
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs:74]: Is it correct to say it's non active if withdrawn date is set or PausedOn date is set?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 29]
```diff
@@ -0,0 +1,295 @@
...
+                processedLearners.Add(uln);
+
+                var learnerCommitments = providerCommitments.CommitmentsForLearner(uln);
+                var earnings = earningsByLearner.EarningsForLearner(uln);
+
+                foreach (var earning in earnings)
+                {
+                    if (earning.HasNonIncentiveEarnings())
+                    {
+                        ProcessEarning(accountsWithNonPayableFlagSet, earning,
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: What does ProcessEarning and why do we have to do it seperately for IncentivePayments?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 30]
```diff
@@ -0,0 +1,295 @@
...
+        private DateTime CalculateCensusDate(RawEarning earning, TransactionTypesFlag earningType)
+        {
+            var date = new DateTime(1900, 01, 01);
+            switch (earningType)
+            {
+                case TransactionTypesFlag.AllLearning:
+                    date = CalculateOnProgCensusDate(earning);
+                    break;
+                case TransactionTypesFlag.FirstEmployerProviderIncentives:
+                    date = earning.FirstIncentiveCensusDate ?? new DateTime(1900,01,01);
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs]: this could be 

date = earn.... ?? date

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 31]
```diff
@@ -0,0 +1,295 @@
...
+        private DateTime CalculateOnProgCensusDate(RawEarning earning)
+        {
+            var month = MonthFromPeriod(earning.Period);
+            var year = YearFromPeriod(earning.Period, earning.EpisodeStartDate ?? new DateTime(9999, 01, 01));
+
+            var lastDayOfMonth = DateTime.DaysInMonth(year, month);
+            return new DateTime(year, month, lastDayOfMonth);
+        }
+
+        private int MonthFromPeriod(int period)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs:161]: We see this code all over the place in payments. I think Period should be elevated to be a common value class.

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs [hunk 32]
```diff
@@ -0,0 +1,295 @@
...
+{
+    public interface IValidateDatalocks
+    {
+        DatalockValidationResult ValidateDatalockForProvider(
+            ProviderCommitments providerCommitments, 
+            IEnumerable<RawEarning> providerEarnings, 
+            ImmutableHashSet<long> accountsWithNonPayableFlagSet);
+    }
+
+    public class DatalockValidationService : IValidateDatalocks
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs:22]: This class is processing by Provider, but there are where clauses inside the code which are filtering by UKPRN, shouldn't the data being passed into this service already do this? It may simplify the logic

File: src/AcceptanceTesting/Features/additional_payments.feature (hunk 1 above)
Comment (carlbm) [reply src/AcceptanceTesting/Features/additional_payments.feature:1110]: The way that the test is structured is unchanged. The second commitment still fails as the price is wrong.
How it was originally is not possible with production data.

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs (hunk 3 above)
Comment (carlbm) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:271]: This is a requirement. If we have a standard, then the framework and pathway codes must be null and the programme type must be 25. 
If we have a framework, then the pathway and programme codes must be populated and the standard code must be null. 
Null and 0 are fairly interchangeable in the tests/code, but always worth checking...

File: src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs (hunk 4 above)
Comment (carlbm) [reply src/AcceptanceTesting/SpecByExample/SFA.DAS.Payments.AcceptanceTests/TableParsers/CommitmentsTableParser.cs:275]: This is a requirement. If we have a standard, then the framework and pathway codes must be null and the programme type must be 25.
If we have a framework, then the pathway and programme codes must be populated and the standard code must be null.

File: src/PeriodEnd/PaymentsDue/DeployPaymentsDue/copy mappings/DasPaymentsDueILRCopyToTransientMapping.xml (hunk 5 above)
Comment (carlbm) [reply src/PeriodEnd/PaymentsDue/DeployPaymentsDue/copy mappings/DasPaymentsDueILRCopyToTransientMapping.xml:1]: Driven by DC. They made a few changes that I've retrofitted

File: src/PeriodEnd/PaymentsDue/DeployPaymentsDue/sql/ddl/Ilr.Bulk.PaymentsDue.DDL.Views.sql (hunk 6 above)
Comment (carlbm) [reply src/PeriodEnd/PaymentsDue/DeployPaymentsDue/sql/ddl/Ilr.Bulk.PaymentsDue.DDL.Views.sql]: Driven by DC. They made a few changes that I've retrofitted

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.IntTests/Infrastructure/Data/Repositories/Temp.cs (hunk 7 above)
Comment (carlbm) [reply src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.IntTests/Infrastructure/Data/Repositories/Temp.cs:9]: If there are no tests in a test assembly the build can fail (not 100% sure) but could remove it...

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/PausedCommitmentMatcher.cs (hunk 10 above)
Comment (carlbm) [reply src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/PausedCommitmentMatcher.cs:15]: Default is true from the base class, so do need to override it

**Review by Pauljgraham — APPROVED** [review]

**Review by Pauljgraham — APPROVED** [review]

**Comment by satya-sanikommu** [comment]
Review and Testing completed. Gave the components to DC and tested in PreProd. Ready to merge
